adagrams/game.py

```diff
@@ -1,11 +1,151 @@
+    'Y': 4, 
+    'Z': 10
+    }
+
+
+def generate_letter_pool():
+    """Return a list of all the available letters."""
+    letter_pool_list = []
+    for letter, quantity in LETTER_POOL.items():
+        letter_pool_list.extend([letter] * quantity)
+    return letter_pool_list
+
+
 def draw_letters():
-    pass
+    """Return a list of 10 randomly chosen letters from the letter pool."""
+    letter_pool_list = generate_letter_pool()
+    letter_bank = []
+    for letter in range(0, 10):
+        random_letter = random.choice(letter_pool_list)
+        letter_pool_list.remove(random_letter)
+        letter_bank.append(random_letter)
+    return letter_bank
+
+
+def count_available_letters(letter_bank):
+    """Return a dictionary containing the quantity of each letter in letter_bank."""
+    letter_counts = {} 
+    for letter in letter_bank:
+        letter_counts[letter] = 0
+    for letter in letter_bank:
+        letter_counts[letter] += 1
+    return letter_counts
+
 
 def uses_available_letters(word, letter_bank):
-    pass
+    """Return True if a word uses only the letters from letter_bank."""
+    letter_counts = count_available_letters(letter_bank)
+    word = word.upper()
+    for letter in word:
+        if letter not in letter_bank or letter_counts[letter] <= 0:
+            return False
+        letter_counts[letter] -= 1
+    return True
+
 
 def score_word(word):
-    pass
+    """Return the amount of points (integer) for a word."""
+    score = 0
+    if len(word) >= WORD_LENGTH_BONUS_THRESHOLD:
+        score += WORD_LENGTH_BONUS_POINTS
+    for letter in word:
+        score += POINT_VALUES[letter.upper()]
+    return score
+
+
+def make_word_and_score_dictionary(word_list):
+    """Score all words and return dictionary with words and scores."""
+    words_and_scores = {}
+    for word in word_list:
+        words_and_scores[word] = score_word(word)
+    return words_and_scores
+
+
+def find_top_score_and_best_words(word_list):
+    """Find the top scoring words and return them along with corresponding score."""
+    words_and_scores = make_word_and_score_dictionary(word_list)
+    words_with_highest_scores = []
+    highest_score = max(words_and_scores.values())
+    for word, score in words_and_scores.items():
+        if score == highest_score:
+            words_with_highest_scores.append(word)
+    return words_with_highest_scores, highest_score
+    
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    """Find the best word out of the words with the top score. Return as tuple with its score."""
+    words_to_compare, highest_score = find_top_score_and_best_words(word_list)
+
+    # Return if only one winner
+    if len(words_to_compare) == 1:
+        return words_to_compare[0], highest_score
+    
+    current_winner = words_to_compare[0]
+    # If there's a tie, the first word to have ten letters wins
+    if len(current_winner) == 10:
+        return current_winner, highest_score
```

Comment:
I'd say wait to use this inside the for loop only, so that we aren't repeating code.